Add plugin: Notes Merger

niffka opened the pull request:
# I am submitting a new Community Plugin

## Repo URL

<!--- Paste a link to your repo here for easy access -->
Link to my plugin: https://github.com/niffka/notes-merger

## Release Checklist
- [x] I have tested the plugin on
  - [x]  Windows
  - [ ]  macOS
  - [x]  Linux (Fedora 37)
  - [x]   ~~Android _(if applicable)_~~
  - [x]  ~~iOS _(if applicable)_~~
- [x] My GitHub release contains all required files
  - [x] `main.js`
  - [x] `manifest.json`
  - [x] `styles.css` _(optional)_
- [x] GitHub release name matches the exact version number specified in my manifest.json (_**Note:** Use the exact version number, don't include a prefix `v`_)
- [x] The `id` in my `manifest.json` matches the `id` in the `community-plugins.json` file.
- [x] My README.md describes the plugin's purpose and provides clear usage instructions.
- [x] I have read the tips in https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines and have self-reviewed my plugin to avoid these common pitfalls.
- [x] I have added a license in the LICENSE file.
- [x] My project respects and is compatible with the original license of any code from other plugins that I'm using.
      I have given proper attribution to these other projects in my `README.md`.

joethei commented:
The screenshot in your readme is missing.

[this.app.workspace.detachLeavesOfType(VIEW_CONTENT_COMPOSE_NOTES);](https://github.com/niffka/notes-merger/blob/584c95dd47c3827c9627b8a2215c6551820841bf/main.ts#L33)
Do not detach leaves with your custom view here, this is an antipattern, see: <https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Don't+detach+leaves+in+%60onunload%60>
Instead, only create a new leaf, if there is not one already.

[this.app.workspace.detachLeavesOfType(VIEW_CONTENT_COMPOSE_NOTES);](https://github.com/niffka/notes-merger/blob/584c95dd47c3827c9627b8a2215c6551820841bf/main.ts#L46)
Also don't detach here.

[containerEl.createEl('h2', {text: 'Settings'});](https://github.com/niffka/notes-merger/blob/584c95dd47c3827c9627b8a2215c6551820841bf/main.ts#L72)
[Only use headings under settings if you have more than one section. ](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Only+use+headings+under+settings+if+you+have+more+than+one+section.)

[const loadedFile = loadedFiles.find((file: TFile) => file.basename === name);](https://github.com/niffka/notes-merger/blob/584c95dd47c3827c9627b8a2215c6551820841bf/src/utils.ts#L9)
[Avoid iterating all files to find a file](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Avoid+iterating+all+files+to+find+a+file+by+its+path)

[const fullPath = (app.vault.adapter as FileSystemAdapter).getFullPath(name);](https://github.com/niffka/notes-merger/blob/584c95dd47c3827c9627b8a2215c6551820841bf/src/utils.ts#L11)

[Prefer the Vault API over the Adapter API](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Prefer+the+Vault+API+over+the+Adapter+APIi)

[const note = readFileSync(absolutePath, { encoding: 'utf8' });](https://github.com/niffka/notes-merger/blob/584c95dd47c3827c9627b8a2215c6551820841bf/src/utils.ts#L17)
Use the Vault API instead, doing so will make your plugin mobile compatible.


[this.app.vault.adapter.write(`${fileName}.md`, this.markdown).then(() => {](https://github.com/niffka/notes-merger/blob/584c95dd47c3827c9627b8a2215c6551820841bf/src/views/generate-markdown.ts#L173)
Prefer async/await over Promise](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Prefer+async%2Fawait+over+Promise)

[contentEl.createEl("h1", { text: "Save As" });](https://github.com/niffka/notes-merger/blob/584c95dd47c3827c9627b8a2215c6551820841bf/src/components/SaveModal.ts#L15)
[Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)

niffka commented:
@joethei Hi, thanks for taking a look at this :) 👍 I have updated the code with the changes from your review. 


> [Avoid iterating all files to find a file](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Avoid+iterating+all+files+to+find+a+file+by+its+path)

I addressed your suggestion, but there's another case (https://github.com/niffka/notes-merger/blob/main/src/views/generate-markdown.ts#L135), where I parse links from a note with regex, and the only way to match them is by iterating through markdown files. I couldn't come up with a better solution, so if there's a better approach, please let me know :)
